content/blog/ebpf-http2-tracing/index.md

```diff
@@ -0,0 +1,145 @@
+
+So if we can’t properly decode HTTP/2 traffic without knowing the state, what can we do?
+
+Fortunately, eBPF technology makes it possible for us to probe into HTTP/2 implementation to get the information that we need, without requiring state.
+
+Specifically, eBPF uprobes address the HPACK issue by directly tracing clear-text data from application memory. By attaching uprobes to the HTTP/2 library APIs that take clear-text headers as input, the uprobes are able to directly read the header content from application memory before they are compressed with HPACK.
+
+[An earlier blog post on eBPF](https://blog.px.dev/ebpf-http-tracing/#tracing-with-uprobes) shows how to implement an uprobe tracer for HTTP applications written in Golang. The first step is to identify the function to attach BPF probes. The function’s arguments need to contain the information we are interested in. The arguments ideally should also have simple structure, such that accessing them in BPF code is easy (through manual pointer chasing). And the function needs to be stable, such that the probes work for a wide range of versions.
+
+Through investigation of the source code of Golang’s gRPC library, we identified `loopyWriter.writeHeader()` as an ideal tracepoint. This function accepts clear text header fields and sends them into the internal buffer. The function signature and the arguments’ type definition is stable, and has not been changed since [2018](https://github.com/grpc/grpc-go/commits/master/internal/transport/controlbuf.go).
+
+Now the challenge is to figure out the memory layout of the data structure, and write the BPF code to read the data at the correct memory addresses.
+
+Let’s take a look at the the signature of the function:
+
+```golang
+func (l *loopyWriter) writeHeader(streamID uint32, endStream bool, hf []hpack.HeaderField, onWrite func())
+```
+
+The task is to read the content of the 3rd argument `hf`, which is a slice of `HeaderField`. We use the `dlv` debugger to figure out the offset of nested data elements, and the results are shown in [`simple-gotracing/http2_trace_uprobe/bpf_program.go`](https://github.com/pixie-io/pixie-demos/blob/main/http2-tracing/uprobe_trace/bpf_program.go).
+
+This code performs 3 tasks:
+
+- [probe_loopy_writer_write_header()]() obtains a pointer to the HeaderField objects held in the slice. A slice resides in memory as a 3-tuple of {pointer, size, capacity}, where the BPF code reads the pointer and size of certain offsets from the SP pointer.
+
+- [submit_headers()]() navigates the list of HeaderField objects through the pointer, by incrementing the pointer with the size of the HeaderField object.
+
+- For each HeaderField object, [copy_header_field()]() copies its content to the output perf buffer. HeaderField is a struct of 2 string objects. Moreover, each string object resides in memory as a 2-tuple of {pointer, size}, where the BPF code copies the corresponding number of bytes from the pointer.
```

Comment:
https://github.com/pixie-io/pixie-demos/blob/34e0b14116a88aa89534dcc1ab8f269a6d520bdd/http2-tracing/uprobe_trace/bpf_program.go#L51

For copy_header_field()